std/machines/hash/keccakf16_memory.asm

```diff
@@ -0,0 +1,876 @@
+use std::array;
+use std::utils;
+use std::utils::unchanged_until;
+use std::utils::force_bool;
+use std::convert::expr;
+use std::convert::int;
+use std::convert::fe;
+use std::prelude::set_hint;
+use std::prelude::Query;
+use std::prover::eval;
+use std::prover::provide_value;
+use std::machines::small_field::memory::Memory;
+use std::machines::small_field::add_sub::AddSub;
+
+machine Keccakf16Memory(mem: Memory, add_sub: AddSub) with
+    latch: first_step,
+    operation_id: operation_id,
+    call_selectors: sel,
+{
+    // ------------- Begin memory read / write ---------------
+    // Additional columns compared to the non-memory version:
+    // - 2 columns (1 high and 1 low) for user input address (of first byte of input).
+    // - 2 columns (1 high and 1 low) for user output address (of first byte of output).
+    // - 98 columns (49 high and 49 low) for computed input/output address of all bytes.
+    // Overall, given that there are 2,600+ columns in the non-memory version, this isn't a huge cost.
+
+    // Methodology description:
+    // 1. User latch in the input and output addresses in the first row of each block.
```

Comment:
```suggestion
    // 1. The latch with the input and output addresses + time step is in the first row of each block.
```
I found this sentence confusing. What's a user latch?

Reply:
Agree and edited. I meant to use "latch" as a verb, so like "user latches in the input and output".